### DISCUSSION TITLE
Fun-out replication in the event of a disaster

### DISCUSSION THREAD

**Discussion by mtopczewski** [zrepl/zrepl#599] [Q&A]
Hi everyone!
 I'm a newbe in zfs and zrepl solution. I wonder if this helps me to solve that case:

I have 2 servers in different cities connected by low speed connection (about 30 Mib). One server is my production server (let's name it as A), the other is backup server (let's name it as B).
I would like to replicate data to B. I tried to run the  "Fun-out replication" scenario from tutorial, but it breakes with error and I can't find the reason. Maybe I missed somthing obvious.

On A I have created zpoll "zfspool_ftp" and mount it to /var/ftp
On B I have created zpool "ftp" and mount it to /var/ftp

Let's ignore the security and the fact that currently on B is less space then on A.

A: 
"zpool list"
NAME          SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
zfspool_ftp   496G   181G   315G        -         -     0%    36%  1.00x    ONLINE  -

"zfs list"
NAME          USED  AVAIL     REFER  MOUNTPOINT
zfspool_ftp   181G   299G      181G  /var/ftp

B:
"zpool list"
NAME     SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
ftp      298G   184G   114G        -         -     1%    61%  1.00x    ONLINE  -

"zfs list"
NAME     USED  AVAIL     REFER  MOUNTPOINT
ftp      184G   104G      182G  /var/ftp

this is zrepl.yml content:

A:
jobs:
- name: snapshots
  type: snap
  filesystems:
    'zfspool_ftp<': true
  snapshotting:
    type: periodic
    prefix: zrepl_
    interval: 10m
  pruning:
    keep:
    - type: regex
      negate: true
      regex: '^zrepl_'
    - type: grid
      grid: 1x1h(keep=all) | 24x1h | 30x1d | 12x30d
      regex: '^zrepl_'

- name: ftp_B
  type: source
  serve:
    type: tcp
    listen: :8888
    clients: {
      "192.168.x.1" : "B.domain.local",
      "192.168.x.2" : "B.domain.local"
    }
  filesystems:
    'zfspool_ftp<': true
  snapshotting:
    type: manual

B:
jobs:
- name: ftp_A
  type: pull
  connect:
    type: tcp
    address: 172.y.x.1:8888
  root_fs: ftp
  interval: 3m
  pruning:
    keep_sender:
    - type: regex
      regex: '.*'
    keep_receiver:
    - type: regex
      negate: true
      regex: '^zrepl_'
    - type: grid
      grid: 1x1h(keep=all) | 24x1h | 30x1d | 12x30d
      regex: '^zrepl_'

Below is screenshot of zrepl status on server B:

![image](https://user-images.githubusercontent.com/44072030/165889235-0ab17d71-1503-4aff-9e0b-5a56c583dcdb.png)

And here are fragment of log with first occurrence of error (command zfs load-key):

Apr 29 07:16:56 B.domain.local zrepl[3901005]: [ftp_GITSRV][trace.data][Uv38$rWXK$rWXK.3Een.BiA1.jJeb.7wRc.mgTg.EMbF]: begin span
Apr 29 07:16:56 B.domain.local zrepl[3901005]: [ftp_GITSRV][zfs.cmd][Uv38$rWXK$rWXK.3Een.BiA1.jJeb.7wRc.mgTg.EMbF]: starting command cmd="zfs load-key"
Apr 29 07:16:56 B.domain.local zrepl[3901005]: [ftp_GITSRV][zfs.cmd][Uv38$rWXK$rWXK.3Een.BiA1.jJeb.7wRc.mgTg.EMbF]: started command cmd="zfs load-key"
Apr 29 07:16:56 B.domain.local zrepl[3901005]: [ftp_GITSRV][zfs.cmd][Uv38$rWXK$rWXK.3Een.BiA1.jJeb.7wRc.mgTg.EMbF]: start waiting cmd="zfs load-key"
Apr 29 07:16:56 B.domain.local zrepl[3901005]: [ftp_GITSRV][zfs.cmd][Uv38$rWXK$rWXK.3Een.BiA1.jJeb.7wRc.mgTg.EMbF]: command exited with error cmd="zfs load-key" usertime_s="0.001124" total_time_s="0.004748152" systemtime_s="0.00332" err="exit status 2"
Apr 29 07:16:56 B.domain.local zrepl[3901005]: [ftp_GITSRV][trace.data][Uv38$rWXK$rWXK.3Een.BiA1.jJeb.7wRc.mgTg.EMbF]: finished span cli.(*Subcommand).run#0$active-side-job#0$active-side-job#0.active-side-job ftp_GITSRV.invocation-1.replication.plan.endpoint.(*Receiver).ListFilesystems.zfs load-key duration_s="0.005021149"

Next errors:

...
command exited with error systemtime_s="0.002749" total_time_s="0.004945924" usertime_s="0.001824" err="exit status 2" cmd="zfs send"

...
command exited with error cmd="zfs send -nvt 1-b95757b5c-d0-789c636064000310a500c4ec50360710e72765a526973030c8ac8328c2904f4b2b4e2d01c9cc80cbb321c9275596a41603e98e987a776cfa4bf2d34b335318189edcf2607fd25463e28024cf0996cf4bcc4d6560a84a2b2ec8cfcf894f2b2970a82a4a2dc88937323032323031b2883734303134348f3730308078830100f4c5200a" total_time_s="0.005227662" systemtime_s="0.002345" usertime_s="0.002367" err="exit status 255"

...
command exited with error usertime_s="0.003089" systemtime_s="0.002074" err="exit status 2" cmd="zfs destroy" total_time_s="0.005537597"

Best regards,
Mariusz
![zrepl_status](https://user-images.githubusercontent.com/44072030/165889803-696a85ed-9215-449b-b657-ca2f5734bf2e.png)

**Comment by problame** [comment]
The error indicates that the snapshot `zfspool_ftp@zrepl_...` doesn't exit on the sender anymore. This appears to be the initial replication. 

The problem is that your `snap` job on A deleted the snapshot before the `pull` job on B could finish transferring the snapshot.

If it were a `push` style setup, I'm certain that it wouldn't happen, due to [the zfs holds that we put in place](https://zrepl.github.io/configuration/replication.html#protection-option).
I have one bug report in the back of my head that complained about this being broken for `pull` setups.

Will try to reproduce your issue and report back.

**Reply by problame** [reply]
Nope, a pull setup will create a step hold on the sending side. That is, from zrepl 0.3 onward. And the status UI you are using suggests that you are using zrepl >= 0.4.

**Reply by problame** [reply]
@InsanePrawn ideas?

**Comment by mtopczewski** [comment]
Additional information:

Server A is Ubuntu 20, and it is VM on VMWare vSphere 5.5: uname -a output: Linux A 5.4.0-109-generic #123-Ubuntu SMP Fri Apr 8 09:10:54 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

Server B is Fedora 35 and it is "real" machine, uname -a output: Linux B.rada.local 5.16.16-200.fc35.x86_64 #1 SMP PREEMPT Sat Mar 19 13:52:41 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

All snapshots on A are available after trying to pull on server B.
